python/pysrc/kaskada/_timestream.py

```diff
@@ -1149,9 +1149,23 @@ def _aggregation(
             raise NotImplementedError(
                 f"Aggregation '{op} does not support trailing windows"
             )
+
         trailing_ns = int(window.duration.total_seconds() * 1e9)
-        # HACK: Use null predicate and number of nanoseconds to encode trailing windows.
-        return Timestream._call(op, input, *args, None, trailing_ns)
+
+        # Create the shifted-forward input
+        input_shift = input.shift_by(window.duration)
+
+        # Merge, then extract just the input.
+        #
+        # Note: Assumes the "input" is discrete. Can probably
+        # use a transform to make it discrete (eg., `input.filter(True)`)
+        # or a special function to do that.
+        #
+        # HACK: This places an extra null row in the input to `collect`
+        # which allows us to "clear" the window when the appropriate 
+        # `duration` has passed with no "real" inputs.
+        merged_input = record({ "input": input, "shift": input_shift }).col("input")
+        return Timestream._call("collect", merged_input, *args, None, trailing_ns)
```

Comment:
Don't we need to do the filtering here to ensure the right number of rows? I guess the answer is no since those rows are the clearing, and we actually want those.

Reply:
Yeah, it's odd. Similar to how a `tick` introduces a row, the `trailing` window introduces a row at the time of the last event + duration. However, this row is really intended to be a "clearing" of the buffer, meaning we shouldn't need it unless a whole `duration` of time elapses. 

It would be easier to manage this behavior if all inputs were known to be "real" inputs -- we could compare the time to the next input and insert a "clearing" row if necessary. 

For now, I think it's okay for our specific use case, but it will introduce some noisy output (doubles the output rows) for larger inputs.